### ISSUE TITLE
Schema for JSON documents describing raster data sets available through the web application.

### ISSUE THREAD

**Issue by tmcphillips** [openskope/skope-deployment#22]
These documents will be indexed by Elasticsearch and the properties in these documents searchable through the web application.

**Comment by jterstriep** [comment]
I have our first version of the dataset metadata. It was originally done as a spreadsheet and with a hybrid data approach where the app would make multiple queries to get the necessary information for workspace. The concept of layers and performing geospatial queries with ES hadn't really gotten there.

For historical purposes only, the original meta data is at https://docs.google.com/spreadsheets/d/1gGsHwRT0gD7YxeGw_bX0VaFpiiiyfC3JRcCT7aL-iV8/edit?usp=sharing

**Comment by abrin** [comment]
@jterstriep is the current one somewhere? I'm curious

**Comment by jterstriep** [comment]
It depends on what you mean by "current". I took a pass at creating a
formal ES mapping definition. It's currently in the elasticsearch directory
of this repo. It is not backwardly compatible with the spreadsheet and we
loaded using csv. I'm not sure that it will live here long-term and it
needs documentation.

**Comment by abrin** [comment]
this [file](https://github.com/openskope/skope-deployment/blob/master/elasticsearch/mapping.json)?

**Comment by jterstriep** [comment]
Yes sir.

**Comment by Zodiase** [comment]
@jterstriep @tmcphillips What's the status on this issue?

**Comment by tmcphillips** [comment]
My understanding from our meetings last week is that I'm handing this off to you, @Zodiase and @jterstriep, now that the test framework for our mapping and searches are set up in [openskope/skope-elasticsearch](https://github.com/openskope/skope-elasticsearch).  

I recommend that the documents to be loaded into Elasticsearch for release 0.1 be maintained in the [data](https://github.com/openskope/skope-elasticsearch/tree/master/data) directory of that repo, the [mapping file ](https://github.com/openskope/skope-elasticsearch/blob/master/skope.dataset.mapping.json )be maintained at the top of the repo, and that the tests continue to exercise both.  This way the mapping and documents can be changed as needed to make the searches work, to return the right info to the meteor app in response to searches, and to describe where to find tiles on GeoServer, and the tests will confirm that documents conform to the schema and that the searches we want to support continue to return the correct results (without having to do many manual tests by hand in the UI each time we change the mapping or change a document to be indexed).

If you like, I can continue to maintain the mapping and tests, but I'll need to know what needs to be changed, what searches we're supporting, etc, so it may be quicker for you to do it yourself.  How would you like to proceed?

**Comment by Zodiase** [comment]
@tmcphillips If you are handing this off then please update the `Assignees` accordingly.

BTW, if you have the ES testing setup ready in that repo, can we somehow clean-up the `openskope/skope-deployment` repo?

I'm not sure what specific things are blocking this issue. But I'm looking at the mapping file and the updated UI in the slides and I can't make the connection between the two. The mapping file also lacks explanation about the purpose of each property, what type of values are expected in each property and how should the values be handled.
Also the slides seem to also contain deprecated design? 

Specific confusions:
- The `datatypes` property is a `keyword` field. But the illustrations in the slides show the "Datatypes" containing complex string values (e.g. `May-September Growing-degree Days (GDD)`). Does that mean the entire string is stored as the "keyword"? What if the UI needs some shorter representations?
- The `period` property is not going to work, as [the documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/range.html) says, `date_range` is

    > A range of date values represented as **unsigned 64-bit integer** milliseconds elapsed since system epoch.

    so my understanding is it can **not** represent any date before `January 1, 1970`.
- It looks like the model data of a model run (dataset) is stored inside a model run (dataset) in the `model` property? What if the model name is changed? Does the `model. version` property mean the model is immutable so existing model info never changes? Is the `model.new` property a link? How does these properties allow the user to get to the model (description) page?

I guess my questions are:
- How to document the mappings better?
- How are we coming up with the mapping? Is it following the UI design (the consumer) or following the source data (the provider), or something else?

**Comment by Zodiase** [comment]
Also it looks like we might need multiple mappings for different indices. How does this mapping file support that?

**Comment by Zodiase** [comment]
Also could somebody quickly explain how the mapping file is loaded? I don't see anything in the docker compose file using it.

**Comment by tmcphillips** [comment]
The mapping that the tests use is the file skope.dataset.mapping.json at the top of the skope-elasticsearch repo.  In that mapping the datatypes property is of type 'text'.  Is this the file your are looking at?  I based this mapping on the contents the 4th tab ("metadata descriptions") of the "SKOPE II Data & Model Sources table" which proposes both the property times and gives a description for each.  We could move this table into a document on GitHub so that it is synchronized with the latest mapping.  There are many changes that need to be made.

My suggestion on fields such as period that won't support the queries we need:  change them to make the searches we need work and include a test that demonstrates that the updated mapping and JSON documents support the searches.

**Comment by tmcphillips** [comment]
The mapping file in skope-elasticsearch is loaded by individual tests (and the index is deleted multiple times).  There should be a different docker compose file that loads the mapping at startup time and provides an elasticsearch instance for use during development (rather than during tests).

**Comment by tmcphillips** [comment]
For multiple indices you can provide different mappings separately, or (I believe) include them in a single PUT request.

**Comment by Zodiase** [comment]
@tmcphillips No I was looking at https://github.com/openskope/skope-elasticsearch/blob/master/mapping.json.

From the `SKOPE II Data & Model Sources Table`, 

> Keywords and probably abbreviations. Comma seperated list

looks to me means short terms instead of long, complex strings.
And `May-September Growing-degree Days (GDD)` doesn't look short to me.
`Net Water-year Precipitation (previous October - current November)` is even longer.

Also I'd not suggest using `text` type as we need some field to aggregate all the data types. `text` type is by default now allowed for aggregation and has horrible performance if enabled. Also I believe the `text` type would make `Net` as an aggregated result, instead of the entire string.

If there is a clear requirement/user story to search with data type by typing in a string, and does not require the ability to filter by selecting from an aggregated list of data types, then we can go with `text` field.

----

I know I can provided multiple mappings separately. I was under the assumption that everything should be defined in one mapping file and somehow that file will be automatically consumed by elastic search.

**Comment by Zodiase** [comment]
@tmcphillips This slide here:

![image](https://user-images.githubusercontent.com/1815174/35018076-8381faea-fae5-11e7-8948-18b189164156.png)

looks to me we still need to aggregate over all the data types. If the search field also needs to search the data types (not by `keyword` but by `text`, meaning matching by stemmed tokens instead of exact string match) then we need the same source info represented in two different properties, one for aggregation (and displayed as a list) and the other for full text searching.

**Comment by tmcphillips** [comment]
I agree.  We may have several properties that are repeated in text (for full text search and display on dataset cards), and as keywords (for filtering, menus, etc).

One column we could add to the metadata descriptions spreadsheet is an indication of whether to include the field in a document-wide full text search field (what used to the _all field but is now built explicitly in the mappings).

**Comment by Zodiase** [comment]
@tmcphillips I think this https://github.com/openskope/SKOPE/issues/14 could be expanded to include all the tasks that need to be supported on the search page?
